File: chapter4/scala/src/main/scala/chapter4/SparkTables.scala
package main.scala.chapter4

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql._

object SparkTables {

  def listTables(ss: SparkSession, tname: String): Unit = {
    ss.catalog.listDatabases().collect().foreach(println(_))
    System.out.println()
    ss.catalog.listTables().collect().foreach(println(_))
    System.out.println()
    ss.catalog.listColumns(tname).collect().foreach(println(_))
  }

  def main(args: Array[String]) {

    val spark = SparkSession
      .builder
      .appName("SparkSQLTablesApp")
      .enableHiveSupport()
      .getOrCreate()

    if (args.length <= 0) {
      println("usage Example4_1 <file path to us_population.json>")
      System.exit(1)
    }
    //get the path to the JSON file
    val json_file = args(0)

    // create database and managed tables
    spark.sql("DROP DATABASE IF EXISTS learn_spark_v2 CASCADE")
    spark.sql("CREATE DATABASE learn_spark_v2")
    spark.sql("USE learn_spark_v2")
    spark.sql("CREATE TABLE us_population_tbl (zipcode STRING, city STRING, loc ARRAY<DOUBLE>, pop INT)")

    //use Catalog to list the database and tables
    System.out.println("Listing Catalog Information")
    listTables(spark, "us_population_tbl")
    // drop the database and create unmanaged tables
    spark.sql("DROP DATABASE IF EXISTS learn_spark_v2 CASCADE")
    spark.sql("CREATE DATABASE learn_spark_v2")
    spark.sql("USE learn_spark_v2")
    spark.sql("CREATE TABLE us_population_tbl(zipcode STRING, city STRING, loc ARRAY<DOUBLE>, pop INT) USING org.apache.spark.sql.json OPTIONS(path '../../data/us_population.json')")

    /**
      * It seems like you cannot bind a variable as an argument in OPTION statement in SQL DDL
      * spark.sql("CREATE TABLE us_population_tbl(zipcode STRING, city STRING, loc ARRAY<DOUBLE>, pop INT) USING org.apache.spark.sql.json OPTIONS(path json_file)")
      **/
    listTables(spark, "us_population_tbl")
  }
}

File: chapter4/py/src/spark_tables.py
from pyspark.sql import SparkSession
import sys
import os.path


def list_tables(ss, tname):
    """
    Use the Catalog API interface to list some databases, tabels, columns
    :param ss: SparkSession
    :return:  none
    """
    print(ss.catalog.listDatabases())
    print()
    print(ss.catalog.listTables())
    print()
    print(ss.catalog.listColumns(tname))

# main program
if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: spark_tables.py <us_population.json>", file=sys.stderr)
        sys.exit(-1)
        # create a SparkSession

    spark = (SparkSession
             .builder
             .appName("SparkSQLTablesApp")
             .enableHiveSupport()
             .getOrCreate())
    # get the path to the JSON file
    json_file = os.path.join(os.getcwd(), sys.argv[1])

    if __name__ == '__main__':
        # create database and managed tables
        spark.sql("DROP DATABASE IF EXISTS learn_spark_v2 CASCADE")
        spark.sql("CREATE DATABASE learn_spark_v2")
        spark.sql("USE learn_spark_v2")
        spark.sql("CREATE TABLE us_population_tbl (zipcode STRING, city STRING, loc ARRAY<DOUBLE>, pop INT)")

        # use Catalog to list the database and tables
        list_tables(spark)
        # drop the database and create unmanaged tables
        spark.sql("DROP DATABASE IF EXISTS learn_spark_v2 CASCADE")
        spark.sql("CREATE DATABASE learn_spark_v2")
        spark.sql("USE learn_spark_v2")
        spark.sql("CREATE TABLE us_population_tbl(zipcode STRING, city STRING, loc ARRAY<DOUBLE>, pop INT) USING org.apache.spark.sql.json OPTIONS(path '../../data/us_population.json')")
        """
        It seems like you cannot bind variables as argument in OPTIONs statement in SQL DDL
        spark.sql("CREATE TABLE us_population_tbl(zipcode STRING, city STRING, loc ARRAY<DOUBLE>, pop INT) USING org.apache.spark.sql.json OPTIONS(path json_file)")
        """
        list_tables(spark)






